photologue/forms.py

```diff
@@ -85,90 +55,12 @@ def clean(self):
-
-            if os.path.dirname(filename):
-                logger.warning('Ignoring file "{}" as it is in a subfolder; all images should be in the top '
-                               'folder of the zip.'.format(filename))
-                if request:
-                    messages.warning(request,
-                                     _('Ignoring file "{filename}" as it is in a subfolder; all images should '
-                                       'be in the top folder of the zip.').format(filename=filename),
-                                     fail_silently=True)
-                continue
-
-            data = zip.read(filename)
-
-            if not len(data):
-                logger.debug('File "{}" is empty.'.format(filename))
-                continue
-
-            photo_title_root = self.cleaned_data['title'] if self.cleaned_data['title'] else gallery.title
-
-            # A photo might already exist with the same slug. So it's somewhat inefficient,
-            # but we loop until we find a slug that's available.
-            while True:
-                photo_title = ' '.join([photo_title_root, str(count)])
-                slug = slugify(photo_title)
-                if Photo.objects.filter(slug=slug).exists():
-                    count += 1
-                    continue
-                break
-
-            photo = Photo(title=photo_title,
-                          slug=slug,
-                          caption=self.cleaned_data['caption'],
-                          is_public=self.cleaned_data['is_public'])
-
-            # Basic check that we have a valid image.
-            try:
-                file = BytesIO(data)
-                opened = Image.open(file)
-                opened.verify()
-            except Exception:
-                # Pillow doesn't recognize it as an image.
-                # If a "bad" file is found we just skip it.
-                # But we do flag this both in the logs and to the user.
-                logger.error('Could not process file "{}" in the .zip archive.'.format(
-                    filename))
-                if request:
-                    messages.warning(request,
-                                     _('Could not process file "{0}" in the .zip archive.').format(
-                                         filename),
-                                     fail_silently=True)
-                continue
-
-            contentfile = ContentFile(data)
-            photo.image.save(filename, contentfile)
-            photo.save()
-            photo.sites.add(current_site)
-            gallery.photos.add(photo)
-            count += 1
-
-        zip.close()
-
-        if request:
-            messages.success(request,
-                             _('The photos have been added to gallery "{0}".').format(
-                                 gallery.title),
-                             fail_silently=True)
+            parse_zip(zip_file, self.cleaned_data['gallery'], self.cleaned_data['title'],
```

Comment:
I'm thinking of the future: we have synchronous mode, we have Celery mode...maybe next year we could extend to another asynchronous service (example RQ). Have lots of 'if' special cases will rapidly become difficult to read. How about each mode has a `parse_zip` function that knows how to process a zip model instance (example: by calling a Celery `delay` task). These functions will all have the same signature. Then at the start of the module we can decide which is the correct function to use according to the settings. Something like:

```
if USE_CELERY:
    zip_handling_method = celery_parse_zip_task
else:
    zip_handling_method = synchronous_parse_zip_task
.
.
.
def save(self, request=None):
    zip_file = self.cleaned_data['zip_file']
    instance.zip_file = zip_file
    instance = super().save(commit=False)
    zip_handling_method(instance.id) # The correct method to call is determined at start of file.
```

Reply:
Makes a lot of sense, I'll change it!